### PR TITLE
Healthcheck url from json

### PR DESCRIPTION
**Pull request by imiro**
## Describe your changes

This PR tries to implement the idea discussed on #1817. Obtain the URL where the jupyter server runs on from a json file located on `/home/$NB_USER/.local/share/jupyter/runtime` directory and use it for healthcheck. Parsing of the json file is done with minimal python currently written as a one-liner command. 

Would appreciate any feedback!

## Issue ticket if applicable

Intended as a solution for https://github.com/jupyter/docker-stacks/issues/1709

## Checklist (especially for first-time contributors)

- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
- [x] I will try not to use force-push to make the review process easier for reviewers
- [ ] I have updated the documentation for significant changes

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (mathbunnyru) [inline .pre-commit-config.yaml:24]: Already merged this change to main branch :)

File: base-notebook/Dockerfile
Comment (mathbunnyru) [inline base-notebook/Dockerfile:73]:
```suggestion
    CMD cat /home/"${NB_USER}"/.local/share/jupyter/runtime/*.json | \
```

File: tests/base-notebook/test_healthcheck.py
Comment (mathbunnyru) [inline tests/base-notebook/test_healthcheck.py:47]: Wow, I like the number of tests you have 👍 

I have two questions though:
1. Why do you need to change workdir?
2. We do `time.sleep(15)`. I think when we have this many tests it's better to return early (for example have a loop for 15 seconds maximum, which checks each 0.1 second, whether the status is healthy).

**Review by mathbunnyru — CHANGES_REQUESTED** [review]

File: tests/base-notebook/test_healthcheck.py
Comment (imiro) [reply tests/base-notebook/test_healthcheck.py:47]: 1. I included it just so it resembles the example on https://jupyter-docker-stacks.readthedocs.io/en/latest/using/common.html#user-related-configurations more closely. It's not really needed actually, I tested without changing the workdir as well and it doesn't cause any problems. Should we not change it to simplify the test?
2. That's a good point. I can try to implement it. When testing locally it didn't feel like it took that long though, I guess pytest runs those tests in parallel

File: base-notebook/Dockerfile
Comment (imiro) [reply base-notebook/Dockerfile:73]: Is this to handle the case where "NB_USER" contains whitespace? I'm wondering if we need to quote the entire path `"/home/${NB_USER}/.local/share/jupyter/runtime/*.json"` instead..?

File: .pre-commit-config.yaml
Comment (imiro) [reply .pre-commit-config.yaml:24]: Thanks! I'll merge the latest update from the main branch here

File: tests/base-notebook/test_healthcheck.py
Comment (mathbunnyru) [reply tests/base-notebook/test_healthcheck.py:47]: 1. I think if it doesn't change behaviour in any visible way, it's better to remove it.

File: base-notebook/Dockerfile
Comment (mathbunnyru) [reply base-notebook/Dockerfile:73]: > Is this to handle the case where "NB_USER" contains whitespace?

Yes, though it is not well tested 😂 
Abd we try to keep this style where we use variables with possible space, so it should probably work fine.

> I'm wondering if we need to quote the entire path "/home/${NB_USER}/.local/share/jupyter/runtime/*.json" instead..?

No, if you quote the entire string, `cat` is gonna think you need one specific file (and it won't find it).
So, it should be done the way I wrote it.

**Comment by mathbunnyru** [comment]
@imiro hi! any chance you could take a look at my suggestions?

**Comment by mathbunnyru** [comment]
@imiro are you still interested in this PR?

**Comment by imiro** [comment]
HI @mathbunnyru sorry for getting this stalled, been struggling to find the time to work on this. I've pushed some updates just now

File: base-notebook/Dockerfile
Comment (imiro) [reply base-notebook/Dockerfile]: Actually I tried to test the current implementation without quoting NB_USER and setting NB_USER to have space and it somehow passed the test 🤔 I committed this anyway though

**Comment by imiro** [comment]
If they look good I'll merge the latest commit from main here, and update the documentation as well

**Review by mathbunnyru — COMMENTED** [review]
LGTM!

Please, merge main and update docs as you wanted 👍

**Comment by imiro** [comment]
image
I was testing locally just now and I got this error. Don't remember seeing this before, it seems that `/home/testu` is not created event though I started the container with `-e NB_USER=testu` ?
This also reveals a problem with the test because its exit status remains 0 .. so the status is "healthy"

**Comment by mathbunnyru** [comment]
@imiro thank you!

I think I know several problems with current solution:

1. Using bash is a bad idea. I suggest you rewrite the HEALTHCHECK command to run a simple python script. This way, if the file doesn't exist or something else, we will have an apparent problem, which won't be skipped during bash settings (like pipefail and so on).

2. When you use `-e NB_USER=testu`, you also have to use `user=root`, otherwise you won't be able to change username.

File: base-notebook/Dockerfile
Comment (mathbunnyru) [inline base-notebook/Dockerfile:73]: This is getting out of hand! 😂 

Let's write a simple python script, please 😄

Comment (imiro) [reply base-notebook/Dockerfile:73]: yeah I know =)) adding a simple bash option seems to resolve the issue, so I thought I'll give it a shot lol. alright, I'll try to work on the python script

Comment (imiro) [reply base-notebook/Dockerfile:73]: Anyway, when I run `make build/base-notebook` followed by `make test/base-notebook`, the test should run on the image I have locally, right?

Comment (imiro) [reply base-notebook/Dockerfile:73]: Actually, never mind. I got strange behavior, apparently because I forgot to rebuild the image locally

Comment (mathbunnyru) [reply base-notebook/Dockerfile:73]: > Anyway, when I run `make build/base-notebook` followed by `make test/base-notebook`, the test should run on the image I have locally, right?

Yes, but you need to build all parent images as well - in this case, if you want to have the latest version, you should first build `make build/docker-stacks-foundation`.

**Comment by mathbunnyru** [comment]
@imiro this works as expected now, right?

**Review by mathbunnyru — APPROVED** [review]

**Comment by imiro** [comment]
@mathbunnyru Just added a couple more things. I was halfway there sorry.

So another observation is that NB_USER values with spaces aren't valid. I modified the test a bit for this. Please review, thanks!

File: tests/base-notebook/test_healthcheck.py
Comment (mathbunnyru) [inline tests/base-notebook/test_healthcheck.py:77]: This should actually work.
I think it's better to fix users with spaces after merging this PR.

Comment (imiro) [reply tests/base-notebook/test_healthcheck.py:77]: Why do you think it should work? I think it's a limitation of Unix in general, you just can't have a username that contain spaces.

Comment (imiro) [reply tests/base-notebook/test_healthcheck.py:77]: I'm not really sure though. That's just my impression when reading the error message.

Comment (imiro) [reply tests/base-notebook/test_healthcheck.py:77]: Sorry you're right. At least based on https://serverfault.com/a/73101 and useradd's help, it should be possible to allow space, by giving the `--badnames` param to the `useradd` command. But it's an entirely different issue I guess, as you said. I'll leave it to you whether to address this issue and actually allow "bad username". 

Should we remove these tests, then?

Comment (mathbunnyru) [reply tests/base-notebook/test_healthcheck.py:77]: You're right.

```
# useradd "crazy user"
useradd: invalid user name 'crazy user'
```

I will delete these tests then (they don't make much sense I suppose).

**Comment by imiro** [comment]
Thanks!

**Comment by mathbunnyru** [comment]
You're welcome. I also modernised a script (just a little bit) and removed the irrelevant information from the docs.
Thank you for the contribution 👍

**Pull request merged**
